metricbeat/helper/metric.go

```diff
@@ -0,0 +1,143 @@
+package helper
+
+import (
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"gopkg.in/yaml.v2"
+	"time"
+)
+
+// Base metric configuration
+type MetricConfig struct {
+	Period string
+}
+
+// Metric specific data
+// This must be defined by each metric
+type Metric struct {
+	Name string
+	Enabled bool
+
+	// Generic Config existing in all metrics
+	BaseConfig MetricConfig
+
+	// Raw metric specific config
+	// This is provided to convert it into Config later
+	RawConfig interface{}
+
+	// Metric specific config
+	Config interface{}
+
+	Metricer Metricer
+	Module   *Module
+}
+
+// Interface for each metric
+type Metricer interface {
+	// TODO: Setup can be optional, so we could make it a second interface we check for?
+	Setup()
+	Fetch() []common.MapStr
+	// TODO: add errors to interface messages
+	// TODO: Add stop method
+}
+
+func NewMetric(name string, metricer Metricer, module *Module) *Metric {
+	return &Metric{
+		Name:     name,
+		Metricer: metricer,
+		Module:   module,
+		Enabled: false,
+	}
+}
+
+func (m *Metric) LoadConfig(config interface{}) {
+
+	bytes, err := yaml.Marshal(m.RawConfig)
+
+	if err != nil {
+		logp.Err("Load metric config error: %v", err)
+	}
+	yaml.Unmarshal(bytes, config)
+}
+
+// Registers metric with module
+func (m *Metric) Register() {
+	m.Module.AddMetric(m)
+}
+
+// RunMetric runs the given metric
+func (m *Metric) Run(b *beat.Beat) {
+
+	if !m.Enabled {
+		logp.Debug("helper", "Not starting metric %s as not enabled.", m.Name)
+		return
+	}
+
+
+	m.Metricer.Setup()
+	period, err := time.ParseDuration(m.BaseConfig.Period)
+
+	if err != nil {
+		logp.Info("Error in parsing period of metric %s: %v", m.Name, err)
+	}
+
+	// If no period set, set default
+	if period == 0 {
+		logp.Info("Setting default period for metric %s as not set.", m.Name)
+		period = 1 * time.Second
+	}
+
+	ticker := time.NewTicker(period)
+	defer ticker.Stop()
+
+	logp.Info("Start metric %s with period %v", m.Name, period)
+
+	for {
+		select {
+		case <-ticker.C:
+		}
+
+		events := m.Metricer.Fetch()
+		newEvents := []common.MapStr{}
+
+		// Default names based on module and metric
+		// These can be overwritten by setting index or / and type in the event
+		indexName := m.Module.Name
+		typeName := m.Name
+		timestamp := common.Time(time.Now())
```

Comment:
Right now in libbeat, the index in the elasticsearch output is defined by the output can can't be overwritten. So at the moment the above code doesn't work yet. 

I'm also not 100% convinced, it is the right concept to follow and we should definitively discuss how and if an event can define the type and the index, or if this is always up to the output.

For our current beats, it works very well that there is one index and multiple types define by the event itself. I'm not sure if that will work for all beats, but especially metricbeat.

One idea I had in mind to make this more flexible and configurable is, that in the output a default index and type can be defined and a field from which the index / type should be automatically taken. So the event itself doesn't know about index and type, but the output will decide where to write it.

As an example for the above assuming there would be a filed module_name and metricset_name in the event:
```
output:
  elasticsearch:
    default_index: metricbeat
    index_filed: module_name
    type_filed: metricset_name
    daily_index_pattern: true # so the date would be added to the index name
```

In the case of filebeat, it could also be as following to make the document_type obsolete

```
output:
  elasticsearch:
    default_index: filebeat
    index_filed: fields.type
    type_filed: metricset_name
```

So far this is just an idea to make index and type more flexible.